File: packages/limen-node/src/cli/keys.ts
import { generateKeyPairSync, randomBytes } from 'node:crypto';
import { generatePrivateKey, privateKeyToAccount } from 'viem/accounts';

export async function runGenerateWebhookSecret(action: string): Promise<void> {
  switch (action) {
    case 'generate-webhook-secret':
      console.log(randomBytes(48).toString('base64'));
      return;

    case 'generate-admin-keypair': {
      const kp = generateKeyPairSync('ed25519', {});
      const pub = kp.publicKey.export({ format: 'der', type: 'spki' }).toString('base64');
      const priv = kp.privateKey.export({ format: 'der', type: 'pkcs8' }).toString('base64');
      console.log(`public:  ${pub}`);
      console.log(`private: ${priv}`);
      return;
    }

    case 'generate-evm-key': {
      const pk = generatePrivateKey();
      const account = privateKeyToAccount(pk);
      console.log(`address:     ${account.address}`);
      console.log(`private_key: ${pk}`);
      console.log('');
      console.log('# Treat the private key like a password.');
      console.log('# For testnet only: fund this address from');
      console.log('#   https://faucet.quicknode.com/base/sepolia   (gas)');
      console.log('#   https://faucet.circle.com/                  (USDC)');
      return;
    }

    default:
      console.error(`unknown action: ${action}`);
      console.error('available: generate-webhook-secret | generate-admin-keypair | generate-evm-key');
      process.exitCode = 1;
  }
}

File: packages/limen-python/limen/cli/keys.py
"""`limen keys` — generate webhook secrets + admin keypairs."""

from __future__ import annotations

import base64
import secrets

import typer

keys_app = typer.Typer(no_args_is_help=True)


@keys_app.command("generate-webhook-secret")
def generate_webhook_secret() -> None:
    typer.echo(base64.b64encode(secrets.token_bytes(48)).decode())


@keys_app.command("generate-admin-keypair")
def generate_admin_keypair() -> None:
    from nacl.signing import SigningKey

    sk = SigningKey.generate()
    vk = sk.verify_key
    typer.echo(f"public:  {base64.b64encode(bytes(vk)).decode()}")
    typer.echo(f"private: {base64.b64encode(bytes(sk)).decode()}")


@keys_app.command("generate-evm-key")
def generate_evm_key() -> None:
    """Generate a fresh EVM (Base / Base Sepolia) signing key."""
    from eth_account import Account  # local import: heavy dep

    acct = Account.create()
    typer.echo(f"address:     {acct.address}")
    typer.echo(f"private_key: 0x{acct.key.hex()}")
    typer.echo("")
    typer.echo("# Treat the private key like a password.")
    typer.echo("# For testnet only: fund this address from")
    typer.echo("#   https://faucet.quicknode.com/base/sepolia   (gas)")
    typer.echo("#   https://faucet.circle.com/                  (USDC)")
